[runtime/corlib] Improve MissingMethodExceptions by including message and signature. Fixes #60505

kumpera opened the pull request:
This PR is missing the following and I'd love if someone could help me here.

Figure out serialization for MethodMissingException - and whether it's ok to break it?
Alternatively, I could use the existing byte[] signature field in MissingMemberException and store a utf8 version of the string currently stored in the new field.

Sort out linker support, this PR adds a dependency to a private constructor so we need the linker to special case it.

marek-safar commented on mcs/class/referencesource/mscorlib/system/missingmethodexception.cs:
```diff
@@ -48,20 +48,24 @@ public override String Message
-                    return Environment.GetResourceString("MissingMethod_Name",
-                                                                       ClassName + "." + MemberName +
-                                                                       (Signature != null ? " " + FormatSignature(Signature) : ""));
+					string res = ClassName + "." + MemberName;
```
this should go under #if MONO

marek-safar commented on mcs/class/referencesource/mscorlib/system/missingmethodexception.cs:
```diff
@@ -48,20 +48,24 @@ public override String Message
     
         // Called from the EE
-        private MissingMethodException(String className, String methodName, byte[] signature)
+        private MissingMethodException(String className, String methodName, String signature, String message) : base (message)
```
You should add a new #if MONO only ctor

marek-safar commented on mcs/class/referencesource/mscorlib/system/missingmethodexception.cs:
```diff
@@ -48,20 +48,24 @@ public override String Message
-                                                                       (Signature != null ? " " + FormatSignature(Signature) : ""));
+					string res = ClassName + "." + MemberName;
+					if (!string.IsNullOrEmpty(signature))
+						res = string.Format (signature, res);
```
This is culture-specific call which should be avoided

kumpera replied:
What's the culture oblivious alternative for string interpolation? The string comes from the runtime.

marek-safar replied:
either don't use it or at least use string.Format with CultureInfo.InvariantCulture

marek-safar commented on mcs/class/referencesource/mscorlib/system/missingmethodexception.cs:
```diff
@@ -48,20 +48,24 @@ public override String Message
+					if (!string.IsNullOrEmpty(signature))
+						res = string.Format (signature, res);
+					else
+						res += "(<UNK>)";
```
I think adding nothing would be better in this case

marek-safar commented on mcs/class/referencesource/mscorlib/system/missingmethodexception.cs:
```diff
@@ -73,5 +77,6 @@ public MissingMethodException(String className, String methodName)
         // If ClassName != null, Message will construct on the fly using it
         // and the other variables. This allows customization of the
         // format depending on the language environment.
+		string signature;
```
I think probably the best solution is [NotSerialized] here

marek-safar commented on mono/utils/mono-error.c:
```diff
@@ -648,7 +660,24 @@ mono_error_prepare_exception (MonoError *oerror, MonoError *error_out)
+					break;
+				}
+			}
+			exception = mono_exception_from_name_four_strings_checked (mono_defaults.corlib, "System", "MissingMethodException", type_name, method_name, signature, message, error_out);
```
You also need to update https://github.com/mono/mono/blob/master/mcs/class/corlib/LinkerDescriptor/mscorlib.xml#L252

kumpera commented:
The new commit fixes all issues highlighted my Marek.

- Wrapped changes in referencesources around `#if mono`
- Added [NonSerializable] to the new field
- Used Invariant Culture with string.Format
- Updated the linker descriptor for corlib

kumpera commented:
We now get a sweet `System.MissingMethodException: void Y.Foo<!0>(!!0)`
